### ISSUE TITLE
cannot run letsencrypt in Ubuntu 14.04.4 x64 

### ISSUE THREAD

**Issue by lawrencehui** [certbot/certbot#2804]
Previously I just run letsencrypt-auto and will be able to create the cert effortlessly. However today I have encounter the follow issue:

Command "/root/.local/share/letsencrypt/bin/python2.7 -u -c "import setuptools, tokenize;**file**='/tmp/pip-build-mTEeBV/cryptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(**file**).read().replace('
', '
'), **file**, 'exec'))" install --record /tmp/pip-M1jUNg-record/install-record.txt --single-version-externally-managed --compile --install-headers /root/.local/share/letsencrypt/include/site/python2.7/cryptography" failed with error code 1 in /tmp/pip-build-mTEeBV/cryptography
/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:120: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
You are using pip version 8.0.3, however version 8.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

anyone can show me some lights on how to upgrade the pip inside the virtualenv? I'm really stuck on this at the moment...

**Comment by robertsky** [comment]
try adding a swapfile file. might work.

**Comment by bmw** [comment]
Your version of `pip` is unlikely to be the source of your problems. For security/stability reasons, `letsencrypt-auto`, installs specific versions of Python packages that we have tested. With that said, if you want to upgrade your `pip` version, run:

```
${XDG_DATA_HOME:-~/.local/share}/letsencrypt/bin/pip install -U pip
```

**Comment by StanBright** [comment]
I just hit the same issue. Unfortunately, I can't give more details, except, it worked correctly on a fresh v14.04.04 installation and didn't on an older v14.04

**Comment by Moonomo** [comment]
@bmw that doesn't work. `~/.local/share}/letsencrypt/bin/pip` are replaced by the `letsencrypt-auto` and gives similar error every time.

**Comment by bmw** [comment]
@Moonomo, you're right. What I posted won't work here.

In this case, `letsencrypt-auto` sees that there is no `letsencrypt` executable and recreates the virtualenv. What I posted only works if you have already installed `letsencrypt` but want to update `pip` in the `virtualenv`.

**Comment by oliverwehn** [comment]
Had the issue on a Digital Ocean droplet. [Creating a swap file](http://www.cyberciti.biz/faq/linux-add-a-swap-file-howto/) worked for me. Thx.

**Comment by vpmedia** [comment]
I can confirm that after rebooting my machine and freeing memory the installation succeeded.
Before that I've encountered the same error messages.

**Comment by littleguga** [comment]
@vpmedia rebooting helped for me.

But it will be great to found solution without rebooting server.

**Comment by vpmedia** [comment]
I think creating a swap file is enough, I needed that reboot anyway (for libc updates).

**Comment by CodingContraption** [comment]
It fails for me too with only ~15% system memory usage. Reboot did not help. Unable to create swap partition. (I run it on a VPS hosted by Strato, I assume they blocked it or something)

**Comment by littleguga** [comment]
@CodingContraption any errors while creating swap file?

**Comment by CodingContraption** [comment]
@littleguga `swapon: /swapfile: swapon failed: Operation not permitted
`

This is run as root user.

**Comment by littleguga** [comment]
@CodingContraption it's problem with OpenVZ, so if you can change virtualization type - it will help. If not - change your VPS provider, e.g. for Digital Ocean.

**Comment by CodingContraption** [comment]
@littleguga The pricing difference is too big for me to do that. ($80 vs €16 for the same amount of ram and cpu, yet 5x the storage and unlimited traffic at Strato.)

Since changing virtualization type is out of my hands, I think I'll just have to accept it won't work for me.

**Comment by GabLeRoux** [comment]
Very similar to #2907, was a memory problem in my case.

**Comment by benishor** [comment]
Thanks for the swap hint! I had the same issue on my digitalocean VPS and that was weird since I was really sure I enabled swap before. What happened was a kernel upgrade which required a reboot. Be sure to make your swap persistent!

**Comment by jrlangford** [comment]
Freeing some memory solved the issue for me, didn't need to reboot or add a swapfile but all your comments pointed me to the right direction 👍

**Comment by bmw** [comment]
This should be fixed in our next release with #4978.
